[CI] DataAndIndexLifecycleMixingTests testUpdateIndexTemplateFromILMtoBothILMAndDLM failing

**Build scan:**
https://gradle-enterprise.elastic.co/s/um3hmnoqxhbbu/tests/:x-pack:plugin:ilm:internalClusterTest/org.elasticsearch.xpack.ilm.DataAndIndexLifecycleMixingTests/testUpdateIndexTemplateFromILMtoBothILMAndDLM

**Reproduction line:**
```
./gradlew ':x-pack:plugin:ilm:internalClusterTest' --tests "org.elasticsearch.xpack.ilm.DataAndIndexLifecycleMixingTests.testUpdateIndexTemplateFromILMtoBothILMAndDLM" -Dtests.seed=6CA280320E5660DD -Dtests.locale=es-AR -Dtests.timezone=Asia/Jayapura -Druntime.java=20
```

**Applicable branches:**
8.9

**Reproduces locally?:**
Didn't try

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.xpack.ilm.DataAndIndexLifecycleMixingTests&tests.test=testUpdateIndexTemplateFromILMtoBothILMAndDLM

**Failure excerpt:**
```
java.util.concurrent.ExecutionException: [.ds-datastream-nklrqotqskgq-2023.08.22-000001] org.elasticsearch.index.IndexNotFoundException: no such index [.ds-datastream-nklrqotqskgq-2023.08.22-000001]

  at __randomizedtesting.SeedInfo.seed([6CA280320E5660DD:14D7078B77D91C70]:0)
  at org.elasticsearch.action.support.PlainActionFuture$Sync.getValue(PlainActionFuture.java:340)
  at org.elasticsearch.action.support.PlainActionFuture$Sync.get(PlainActionFuture.java:327)
  at org.elasticsearch.action.support.PlainActionFuture.get(PlainActionFuture.java:95)
  at org.elasticsearch.xpack.ilm.DataAndIndexLifecycleMixingTests.lambda$testUpdateIndexTemplateFromILMtoBothILMAndDLM$7(DataAndIndexLifecycleMixingTests.java:309)
  at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:1160)
  at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:1145)
  at org.elasticsearch.xpack.ilm.DataAndIndexLifecycleMixingTests.testUpdateIndexTemplateFromILMtoBothILMAndDLM(DataAndIndexLifecycleMixingTests.java:305)
  at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
  at java.lang.reflect.Method.invoke(Method.java:578)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:1623)

  Caused by: org.elasticsearch.index.IndexNotFoundException: no such index [.ds-datastream-nklrqotqskgq-2023.08.22-000001]

    at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver.notFoundException(IndexNameExpressionResolver.java:449)
    at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver$ExplicitResourceNameFilter.ensureAliasOrIndexExists(IndexNameExpressionResolver.java:1575)
    at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver$ExplicitResourceNameFilter.filterUnavailable(IndexNameExpressionResolver.java:1555)
    at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver.resolveExpressions(IndexNameExpressionResolver.java:262)
    at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver.concreteIndices(IndexNameExpressionResolver.java:337)
    at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver.concreteIndexNames(IndexNameExpressionResolver.java:328)
    at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver.concreteIndexNames(IndexNameExpressionResolver.java:87)
    at org.elasticsearch.action.support.master.info.TransportClusterInfoAction.checkBlock(TransportClusterInfoAction.java:53)
    at org.elasticsearch.action.support.master.info.TransportClusterInfoAction.checkBlock(TransportClusterInfoAction.java:24)
    at org.elasticsearch.action.support.master.TransportMasterNodeAction.checkBlockIfStateRecovered(TransportMasterNodeAction.java:135)
    at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction.doStart(TransportMasterNodeAction.java:199)
    at org.elasticsearch.action.support.master.TransportMasterNodeAction.doExecute(TransportMasterNodeAction.java:171)
    at org.elasticsearch.action.support.master.TransportMasterNodeAction.doExecute(TransportMasterNodeAction.java:53)
    at org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:86)
    at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:61)
    at org.elasticsearch.tasks.TaskManager.registerAndExecute(TaskManager.java:199)
    at org.elasticsearch.client.internal.node.NodeClient.executeLocally(NodeClient.java:112)
    at org.elasticsearch.client.internal.node.NodeClient.doExecute(NodeClient.java:90)
    at org.elasticsearch.client.internal.support.AbstractClient.execute(AbstractClient.java:379)
    at org.elasticsearch.client.internal.FilterClient.doExecute(FilterClient.java:57)
    at org.elasticsearch.client.internal.support.AbstractClient.execute(AbstractClient.java:379)
    at org.elasticsearch.client.internal.support.AbstractClient.execute(AbstractClient.java:365)
    at org.elasticsearch.xpack.ilm.DataAndIndexLifecycleMixingTests.lambda$testUpdateIndexTemplateFromILMtoBothILMAndDLM$7(DataAndIndexLifecycleMixingTests.java:309)
    at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:1160)
    at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:1145)
    at org.elasticsearch.xpack.ilm.DataAndIndexLifecycleMixingTests.testUpdateIndexTemplateFromILMtoBothILMAndDLM(DataAndIndexLifecycleMixingTests.java:305)
    at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
    at java.lang.reflect.Method.invoke(Method.java:578)
    at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
    at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
    at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
    at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
    at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
    at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
    at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
    at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
    at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
    at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
    at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
    at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
    at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
    at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
    at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
    at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
    at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
    at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
    at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
    at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
    at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
    at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
    at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
    at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
    at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
    at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
    at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
    at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
    at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
    at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
    at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
    at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
    at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
    at java.lang.Thread.run(Thread.java:1623)

```

--- a/x-pack/plugin/ilm/src/internalClusterTest/java/org/elasticsearch/xpack/ilm/DataStreamAndIndexLifecycleMixingTests.java
+++ b/x-pack/plugin/ilm/src/internalClusterTest/java/org/elasticsearch/xpack/ilm/DataStreamAndIndexLifecycleMixingTests.java
@@ -416,7 +416,6 @@ public class DataStreamAndIndexLifecycleMixingTests extends ESIntegTestCase {
             List<String> backingIndices = getBackingIndices(dataStreamName);
             String firstGenerationIndex = backingIndices.get(0);
             String secondGenerationIndex = backingIndices.get(1);
-            ;
             String thirdGenerationIndex = backingIndices.get(2);
             String writeIndex = backingIndices.get(3);
             ExplainLifecycleRequest explainRequest = new ExplainLifecycleRequest().indices(

--- a/x-pack/plugin/ilm/src/internalClusterTest/java/org/elasticsearch/xpack/ilm/DataStreamAndIndexLifecycleMixingTests.java
+++ b/x-pack/plugin/ilm/src/internalClusterTest/java/org/elasticsearch/xpack/ilm/DataStreamAndIndexLifecycleMixingTests.java
@@ -57,7 +57,6 @@ import java.util.List;
 import java.util.Locale;
 import java.util.Map;
 
-import static org.elasticsearch.cluster.metadata.DataStream.getDefaultBackingIndexName;
 import static org.elasticsearch.cluster.metadata.DataStreamTestHelper.backingIndexEqualTo;
 import static org.elasticsearch.cluster.metadata.MetadataIndexTemplateService.DEFAULT_TIMESTAMP_FIELD;
 import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
@@ -139,8 +138,9 @@ public class DataStreamAndIndexLifecycleMixingTests extends ESIntegTestCase {
 
         // data stream was rolled over and has 2 indices managed by ILM
         assertBusy(() -> {
-            String firstGenerationIndex = getDefaultBackingIndexName(dataStreamName, 1);
-            String secondGenerationIndex = getDefaultBackingIndexName(dataStreamName, 2);
+            List<String> backingIndices = getBackingIndices(dataStreamName);
+            String firstGenerationIndex = backingIndices.get(0);
+            String secondGenerationIndex = backingIndices.get(1);
             ExplainLifecycleRequest explainRequest = new ExplainLifecycleRequest().indices(firstGenerationIndex, secondGenerationIndex);
             ExplainLifecycleResponse explainResponse = client().execute(ExplainLifecycleAction.INSTANCE, explainRequest).get();
 
@@ -179,9 +179,10 @@ public class DataStreamAndIndexLifecycleMixingTests extends ESIntegTestCase {
         });
 
         assertBusy(() -> {
-            String firstGenerationIndex = getDefaultBackingIndexName(dataStreamName, 1);
-            String secondGenerationIndex = getDefaultBackingIndexName(dataStreamName, 2);
-            String writeIndex = getDefaultBackingIndexName(dataStreamName, 3);
+            List<String> backingIndices = getBackingIndices(dataStreamName);
+            String firstGenerationIndex = backingIndices.get(0);
+            String secondGenerationIndex = backingIndices.get(1);
+            String writeIndex = backingIndices.get(2);
             ExplainLifecycleRequest explainRequest = new ExplainLifecycleRequest().indices(
                 firstGenerationIndex,
                 secondGenerationIndex,
@@ -235,10 +236,11 @@ public class DataStreamAndIndexLifecycleMixingTests extends ESIntegTestCase {
         );
 
         assertBusy(() -> {
-            String firstGenerationIndex = getDefaultBackingIndexName(dataStreamName, 1);
-            String secondGenerationIndex = getDefaultBackingIndexName(dataStreamName, 2);
-            String thirdGenerationIndex = getDefaultBackingIndexName(dataStreamName, 3);
-            String writeIndex = getDefaultBackingIndexName(dataStreamName, 4);
+            List<String> backingIndices = getBackingIndices(dataStreamName);
+            String firstGenerationIndex = backingIndices.get(0);
+            String secondGenerationIndex = backingIndices.get(1);
+            String thirdGenerationIndex = backingIndices.get(2);
+            String writeIndex = backingIndices.get(3);
             ExplainLifecycleRequest explainRequest = new ExplainLifecycleRequest().indices(
                 firstGenerationIndex,
                 secondGenerationIndex,
@@ -296,7 +298,6 @@ public class DataStreamAndIndexLifecycleMixingTests extends ESIntegTestCase {
         client().execute(CreateDataStreamAction.INSTANCE, createDataStreamRequest).get();
 
         indexDocs(dataStreamName, 2);
-
         assertBusy(() -> {
             GetDataStreamAction.Request getDataStreamRequest = new GetDataStreamAction.Request(new String[] { dataStreamName });
             GetDataStreamAction.Response getDataStreamResponse = client().execute(GetDataStreamAction.INSTANCE, getDataStreamRequest)
@@ -306,11 +307,11 @@ public class DataStreamAndIndexLifecycleMixingTests extends ESIntegTestCase {
             List<Index> backingIndices = getDataStreamResponse.getDataStreams().get(0).getDataStream().getIndices();
             assertThat(backingIndices.size(), equalTo(2));
         });
-
         // data stream was rolled over and has 2 indices managed by ILM
         assertBusy(() -> {
-            String firstGenerationIndex = getDefaultBackingIndexName(dataStreamName, 1);
-            String secondGenerationIndex = getDefaultBackingIndexName(dataStreamName, 2);
+            List<String> backingIndices = getBackingIndices(dataStreamName);
+            String firstGenerationIndex = backingIndices.get(0);
+            String secondGenerationIndex = backingIndices.get(1);
             ExplainLifecycleRequest explainRequest = new ExplainLifecycleRequest().indices(firstGenerationIndex, secondGenerationIndex);
             ExplainLifecycleResponse explainResponse = client().execute(ExplainLifecycleAction.INSTANCE, explainRequest).get();
 
@@ -356,11 +357,11 @@ public class DataStreamAndIndexLifecycleMixingTests extends ESIntegTestCase {
             assertThat(getDataStreamResponse.getDataStreams().get(0).getDataStream().getIndices().size(), is(3));
 
         });
-
         assertBusy(() -> {
-            String firstGenerationIndex = getDefaultBackingIndexName(dataStreamName, 1);
-            String secondGenerationIndex = getDefaultBackingIndexName(dataStreamName, 2);
-            String thirdGenerationIndex = getDefaultBackingIndexName(dataStreamName, 3);
+            List<String> backingIndices = getBackingIndices(dataStreamName);
+            String firstGenerationIndex = backingIndices.get(0);
+            String secondGenerationIndex = backingIndices.get(1);
+            String thirdGenerationIndex = backingIndices.get(2);
             ExplainLifecycleRequest explainRequest = new ExplainLifecycleRequest().indices(
                 firstGenerationIndex,
                 secondGenerationIndex,
@@ -382,7 +383,6 @@ public class DataStreamAndIndexLifecycleMixingTests extends ESIntegTestCase {
             assertThat(thirdGenerationExplain.managedByILM(), is(true));
             assertThat(thirdGenerationExplain.getStep(), is(WaitForRolloverReadyStep.NAME));
         });
-
         // let's migrate this data stream to use the data stream lifecycle starting with the next generation
         client().execute(
             PutDataStreamLifecycleAction.INSTANCE,
@@ -413,10 +413,12 @@ public class DataStreamAndIndexLifecycleMixingTests extends ESIntegTestCase {
         });
 
         assertBusy(() -> {
-            String firstGenerationIndex = getDefaultBackingIndexName(dataStreamName, 1);
-            String secondGenerationIndex = getDefaultBackingIndexName(dataStreamName, 2);
-            String thirdGenerationIndex = getDefaultBackingIndexName(dataStreamName, 3);
-            String writeIndex = getDefaultBackingIndexName(dataStreamName, 4);
+            List<String> backingIndices = getBackingIndices(dataStreamName);
+            String firstGenerationIndex = backingIndices.get(0);
+            String secondGenerationIndex = backingIndices.get(1);
+            ;
+            String thirdGenerationIndex = backingIndices.get(2);
+            String writeIndex = backingIndices.get(3);
             ExplainLifecycleRequest explainRequest = new ExplainLifecycleRequest().indices(
                 firstGenerationIndex,
                 secondGenerationIndex,
@@ -485,8 +487,9 @@ public class DataStreamAndIndexLifecycleMixingTests extends ESIntegTestCase {
 
         // data stream was rolled over and has 2 indices managed by ILM
         assertBusy(() -> {
-            String firstGenerationIndex = getDefaultBackingIndexName(dataStreamName, 1);
-            String secondGenerationIndex = getDefaultBackingIndexName(dataStreamName, 2);
+            List<String> backingIndices = getBackingIndices(dataStreamName);
+            String firstGenerationIndex = backingIndices.get(0);
+            String secondGenerationIndex = backingIndices.get(1);
             ExplainLifecycleRequest explainRequest = new ExplainLifecycleRequest().indices(firstGenerationIndex, secondGenerationIndex);
             ExplainLifecycleResponse explainResponse = client().execute(ExplainLifecycleAction.INSTANCE, explainRequest).get();
 
@@ -528,9 +531,10 @@ public class DataStreamAndIndexLifecycleMixingTests extends ESIntegTestCase {
         });
 
         assertBusy(() -> {
-            String firstGenerationIndex = getDefaultBackingIndexName(dataStreamName, 1);
-            String secondGenerationIndex = getDefaultBackingIndexName(dataStreamName, 2);
-            String writeIndex = getDefaultBackingIndexName(dataStreamName, 3);
+            List<String> backingIndices = getBackingIndices(dataStreamName);
+            String firstGenerationIndex = backingIndices.get(0);
+            String secondGenerationIndex = backingIndices.get(1);
+            String writeIndex = backingIndices.get(2);
             ExplainLifecycleRequest explainRequest = new ExplainLifecycleRequest().indices(
                 firstGenerationIndex,
                 secondGenerationIndex,
@@ -581,10 +585,11 @@ public class DataStreamAndIndexLifecycleMixingTests extends ESIntegTestCase {
         });
 
         assertBusy(() -> {
-            String firstGenerationIndex = getDefaultBackingIndexName(dataStreamName, 1);
-            String secondGenerationIndex = getDefaultBackingIndexName(dataStreamName, 2);
-            String thirdGenerationIndex = getDefaultBackingIndexName(dataStreamName, 3);
-            String writeIndex = getDefaultBackingIndexName(dataStreamName, 4);
+            List<String> backingIndices = getBackingIndices(dataStreamName);
+            String firstGenerationIndex = backingIndices.get(0);
+            String secondGenerationIndex = backingIndices.get(1);
+            String thirdGenerationIndex = backingIndices.get(2);
+            String writeIndex = backingIndices.get(3);
             ExplainLifecycleRequest explainRequest = new ExplainLifecycleRequest().indices(
                 firstGenerationIndex,
                 secondGenerationIndex,
@@ -631,10 +636,11 @@ public class DataStreamAndIndexLifecycleMixingTests extends ESIntegTestCase {
 
         // At this point the data stream is not managed by the data stream lifecycle and all the indices fallback to ILM.
         assertBusy(() -> {
-            String firstGenerationIndex = getDefaultBackingIndexName(dataStreamName, 1);
-            String secondGenerationIndex = getDefaultBackingIndexName(dataStreamName, 2);
-            String thirdGenerationIndex = getDefaultBackingIndexName(dataStreamName, 3);
-            String writeIndex = getDefaultBackingIndexName(dataStreamName, 4);
+            List<String> backingIndices = getBackingIndices(dataStreamName);
+            String firstGenerationIndex = backingIndices.get(0);
+            String secondGenerationIndex = backingIndices.get(1);
+            String thirdGenerationIndex = backingIndices.get(2);
+            String writeIndex = backingIndices.get(3);
             ExplainLifecycleRequest explainRequest = new ExplainLifecycleRequest().indices(
                 firstGenerationIndex,
                 secondGenerationIndex,
@@ -684,8 +690,9 @@ public class DataStreamAndIndexLifecycleMixingTests extends ESIntegTestCase {
         });
 
         assertBusy(() -> {
-            String firstGenerationIndex = getDefaultBackingIndexName(dataStreamName, 1);
-            String writeIndex = getDefaultBackingIndexName(dataStreamName, 2);
+            List<String> backingIndices = getBackingIndices(dataStreamName);
+            String firstGenerationIndex = backingIndices.get(0);
+            String writeIndex = backingIndices.get(1);
 
             // let's check the indices are managed by the data stream lifecycle
             ExplainDataStreamLifecycleAction.Response dataStreamLifecycleExplainResponse = client().execute(
@@ -736,9 +743,10 @@ public class DataStreamAndIndexLifecycleMixingTests extends ESIntegTestCase {
         });
 
         assertBusy(() -> {
-            String firstGenerationIndex = getDefaultBackingIndexName(dataStreamName, 1);
-            String secondGenerationIndex = getDefaultBackingIndexName(dataStreamName, 2);
-            String writeIndex = getDefaultBackingIndexName(dataStreamName, 3);
+            List<String> backingIndices = getBackingIndices(dataStreamName);
+            String firstGenerationIndex = backingIndices.get(0);
+            String secondGenerationIndex = backingIndices.get(1);
+            String writeIndex = backingIndices.get(2);
 
             // let's check the previous indices are managed by the data stream lifecycle
             ExplainDataStreamLifecycleAction.Response dataStreamLifecycleExplainResponse = client().execute(
@@ -772,8 +780,9 @@ public class DataStreamAndIndexLifecycleMixingTests extends ESIntegTestCase {
 
         // the new write index is also managed by ILM
         assertBusy(() -> {
-            String thirdGenerationIndex = getDefaultBackingIndexName(dataStreamName, 3);
-            String writeIndex = getDefaultBackingIndexName(dataStreamName, 4);
+            List<String> backingIndices = getBackingIndices(dataStreamName);
+            String thirdGenerationIndex = backingIndices.get(2);
+            String writeIndex = backingIndices.get(3);
 
             ExplainLifecycleRequest explainRequest = new ExplainLifecycleRequest().indices(thirdGenerationIndex, writeIndex);
             ExplainLifecycleResponse explainResponse = client().execute(ExplainLifecycleAction.INSTANCE, explainRequest).get();
@@ -836,5 +845,12 @@ public class DataStreamAndIndexLifecycleMixingTests extends ESIntegTestCase {
 
     private static DataStreamLifecycle customEnabledLifecycle() {
         return DataStreamLifecycle.newBuilder().dataRetention(TimeValue.timeValueMillis(randomMillisUpToYear9999())).build();
+    }
+
+    private List<String> getBackingIndices(String dataStreamName) {
+        GetDataStreamAction.Request getDataStreamRequest = new GetDataStreamAction.Request(new String[] { dataStreamName });
+        GetDataStreamAction.Response getDataStreamResponse = client().execute(GetDataStreamAction.INSTANCE, getDataStreamRequest)
+            .actionGet();
+        return getDataStreamResponse.getDataStreams().get(0).getDataStream().getIndices().stream().map(Index::getName).toList();
     }
 }

